### PR TITLE
fix: error in @aws-sdk/client-sqs when SendMessageBatch entries have undefined attributes

### PR DESCRIPTION
**Pull request by david-luna**
Kudos to @peterabbott to find the issue and give context to start with the fix.

This PR adds a check when instrumenting `SendMessageBatch` command. The `entries` may not have the property `MessageAttributes` or it could be undefined breaking the propagation of traceContext headers.

We also leverage this fix to bump the version of local stack to `v3.0.0`. This new version supports the JSON protocol so we can re-enable tests for `@aws-sdk/client-sqs@3.446.0` and up.

Fixes: #3746
Closes: https://github.com/elastic/apm-agent-nodejs/issues/3719

~NOTE: we will re-enable test for other modules (`aws-sdk`, ..) in a follow up PR.~

### Checklist

- [x] Implement code
- [x] Add tests
- [ ] Update TypeScript typings
- [ ] Update documentation
- [ ] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

File: .tav.yml [hunk 1]
```diff
@@ -593,7 +593,7 @@ aws-sdk:
   # v3.446.0 and v3.447.0 (latest at 2023-11-09) switched the internal format from XML to JSON
   # and localstack does not support it yet. We stop testing at v3.445.0 intil JSON is supported
   # https://github.com/elastic/apm-agent-nodejs/issues/3716
-  versions: '3.15.0 || 3.67.0 || 3.179.0 || 3.261.0 || 3.342.0 || 3.436.0 || 3.445.0'
+  versions: '3.15.0 || 3.67.0 || 3.179.0 || 3.261.0 || 3.342.0 || 3.436.0 || ^3.445.0'
```
Comment (david-luna) [inline .tav.yml:596]: Note for reviewer: with the update of `localstack` image we can now put the caret back

**Review by david-luna — COMMENTED** [review]
/test tav @aws-sdk/client-sqs

**Review by david-luna — COMMENTED** [review]
/test tav aws-sdk,@aws-sdk/client-s3,@aws-sdk/client-sns,@aws-sdk/client-sqs,@aws-sdk/client-dynamodb

**Review by david-luna — COMMENTED** [review]
/test tav aws-sdk

**Comment by david-luna** [comment]
TAV test for `aws-sdk` have fail but I cannot repro locally. In SQS tests the agent is capturing an error

```
docker-node_tests-1  |   {
docker-node_tests-1  |     error: {
docker-node_tests-1  |       id: 'bb4a2f4dc6c7de65c4ef7bd5640abdfc',
docker-node_tests-1  |       timestamp: 1700669489620000,
docker-node_tests-1  |       parent_id: '66dd98b53f7bdce4',
docker-node_tests-1  |       trace_id: 'ed86458ca1298e5c48a31b5edea7463a',
docker-node_tests-1  |       transaction_id: '2d4db27f74e4d1c9',
docker-node_tests-1  |       transaction: [Object],
docker-node_tests-1  |       context: [Object],
docker-node_tests-1  |       exception: [Object],
docker-node_tests-1  |       culprit: 'ENOTFOUND_ERROR (/app/node_modules/aws-sdk/lib/event_listeners.js)'
docker-node_tests-1  |     }
docker-node_tests-1  |   }
```

**Review by david-luna — COMMENTED** [review]
/test tav aws-sdk

File: test/instrumentation/modules/aws-sdk/sqs.test.js [hunk 2]
```diff
@@ -246,6 +246,7 @@ tape.test('SQS usage scenario', function (t) {
         });
 
         // First the transaction.
+        console.log(JSON.stringify(events, null, 2));
```
Comment (trentm) [inline test/instrumentation/modules/aws-sdk/sqs.test.js:249]: Is this debug logging to review?

File: test/instrumentation/modules/aws-sdk/s3.test.js [hunk 3]
```diff
@@ -228,7 +228,7 @@ tape.test('simple S3 usage scenario', function (t) {
                   resource: 's3/elasticapmtest-bucket-1',
                 },
               },
-              http: { status_code: 200, response: { encoded_body_size: 58 } },
+              http: { status_code: 200 },
```
Comment (trentm) [inline test/instrumentation/modules/aws-sdk/s3.test.js:231]: Interesting. I vaguely recall seeing changes like this before. I don't recall why the change.

**Review by trentm — APPROVED** [review]

File: test/instrumentation/modules/aws-sdk/sqs.test.js (hunk 2 above)
Comment (trentm) [reply test/instrumentation/modules/aws-sdk/sqs.test.js:249]: Oh, this is for debugging the TAV failure.

**Comment by david-luna** [comment]
Printing the stack of the error points to a part of `aws-sdk` handling connection error

```js
if (err.code === 'NetworkingError' && isDNSError(err)) {
    var message = 'Inaccessible host: `' + err.hostname + '\' at port `' + err.port +
      '\'. This service may not be available in the `' + err.region +
      '\' region.';
    this.response.error = AWS.util.error(new Error(message), {
      code: 'UnknownEndpoint',
      region: err.region,
      hostname: err.hostname,
      retryable: true,
      originalError: err
    });
}
```

**Review by david-luna — COMMENTED** [review]
/test tav aws-sdk

**Comment by trentm** [comment]
Notes on understanding the DNS failure when running in CI:

```
docker-node_tests-1  |       exception: {
docker-node_tests-1  |         message: "Inaccessible host: `sqs.us-east-2.localhost.localstack.cloud' at port `4566'. This service may not be available in the `us-east-2' region.",
docker-node_tests-1  |         type: 'UnknownEndpoint',
```

Running localstack:3.0.0 locally:

```
% docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED              STATUS                   PORTS                                             NAMES
1b05233dc3f5   alpine                          "/bin/sh"                About a minute ago   Up About a minute                                                          determined_noether
2ddcd172bedc   localstack/localstack:3.0.0     "docker-entrypoint.sh"   7 minutes ago        Up 7 minutes (healthy)   4510-4559/tcp, 5678/tcp, 0.0.0.0:4566->4566/tcp   docker-localstack-1
...
```

Per https://docs.localstack.cloud/references/configuration/#core the default `LOCALSTACK_HOST` is `localhost.localstack.cloud:4566`, at least in v3. (Was it `localstack:4566` or otherwise in v2?)

From a terminal outside of Docker, I can reach that endpoint and `sqs[.region].`-prefixed endpoints:

```
% curl -i http://localhost.localstack.cloud:4566/
HTTP/1.1 200
Content-Type: text/plain; charset=utf-8
Connection: close
Content-Length: 0
date: Wed, 22 Nov 2023 19:25:51 GMT
server: hypercorn-h11

% curl -i http://sqs.localhost.localstack.cloud:4566/
HTTP/1.1 500
Content-Type: text/xml
Content-Length: 403
Connection: close
date: Wed, 22 Nov 2023 19:25:54 GMT
server: hypercorn-h11

<?xml version='1.0' encoding='utf-8'?>
<ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Code>InternalError</Code><Message>exception while calling sqs-query with unknown operation: Operation detection failed. Missing Action in request for query-protocol service ServiceModel(sqs-query).</Message></Error><RequestId>345c2394-639e-465f-81cc-c11ca5f89eaf</RequestId></ErrorResponse>%

% curl -i http://sqs.us-east-2.localhost.localstack.cloud:4566/
HTTP/1.1 500
Content-Type: text/xml
Content-Length: 403
Connection: close
date: Wed, 22 Nov 2023 19:26:00 GMT
server: hypercorn-h11

<?xml version='1.0' encoding='utf-8'?>
<ErrorResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><Error><Code>InternalError</Code><Message>exception while calling sqs-query with unknown operation: Operation detection failed. Missing Action in request for query-protocol service ServiceModel(sqs-query).</Message></Error><RequestId>030115be-6d1c-4f7b-a8e7-bd6c381a59fc</RequestId></ErrorResponse>%
```


From inside docker I can ping `localstack` and `localhost.localstack.cloud`

```
% docker run -ti --rm --network docker_default alpine /bin/sh
/ # ping localstack
PING localstack (172.26.0.2): 56 data bytes
64 bytes from 172.26.0.2: seq=0 ttl=64 time=0.083 ms
64 bytes from 172.26.0.2: seq=1 ttl=64 time=0.072 ms
64 bytes from 172.26.0.2: seq=2 ttl=64 time=0.246 ms
^C
--- localstack ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.072/0.133/0.246 ms
/ # ping localhost.localstack.cloud
PING localhost.localstack.cloud (127.0.0.1): 56 data bytes
64 bytes from 127.0.0.1: seq=0 ttl=64 time=0.032 ms
64 bytes from 127.0.0.1: seq=1 ttl=64 time=0.063 ms
^C
```

but not accessing the endpoint only works for `localstack`:

```
/ # curl -i http://localhost.localstack.cloud:4566/
curl: (7) Failed to connect to localhost.localstack.cloud port 4566 after 75 ms: Couldn't connect to server
/ # curl -i http://localstack:4566/
HTTP/1.1 200
Content-Type: text/plain; charset=utf-8
Connection: close
Content-Length: 0
date: Wed, 22 Nov 2023 19:21:26 GMT
server: hypercorn-h11

/ #
```

**Comment by trentm** [comment]
Repeating that commit message:

```
IIUC, since localstack/localstack:2.3.0 the default LOCALSTACK_HOST
is `localhost.localstack.cloud:4566` which relies on *external* DNS
to resolve `*.localhost.localstack.cloud` hostnames -- as real AWS
server and client tools do for some services like S3 and SQS.

We can't rely on this external DNS access for our in-docker tests,
i.e. for the current TAV tests. We avoid using this with our
LOCALSTACK_HOST setting and tell the (apparently server-side) handling
for SQS queue URLs to not use hostname-based queue URLs.

This is similar to path vs host style for S3, which is apparently
client-side determined -- again IIUC from
https://docs.localstack.cloud/user-guide/aws/s3/#path-style-and-virtual-hosted-style-requests
```


TODO:
- [ ] update the .tav.yml range for aws-sdk. I think we should be able to use the more recent versions that use the JSON API format for SQS.

**Review by trentm — COMMENTED** [review]
/test tav aws-sdk,@aws-sdk/client-sqs

**Comment by trentm** [comment]
> NOTE: we will re-enable test for other modules (`aws-sdk`, ..) in a follow up PR.

Oh! I went ahead and did it for others. :)

**Review by trentm — COMMENTED** [review]
/test tav aws-sdk,@aws-sdk/client-s3,@aws-sdk/client-sns,@aws-sdk/client-sqs,@aws-sdk/client-dynamodb


**Update:**: These all passed: https://github.com/elastic/apm-agent-nodejs/actions/runs/6964070611/job/18950736206

**Review by trentm — DISMISSED** [review]
@david-luna You'll also need to approve for merging, because the current last commit is mine.

**Comment by david-luna** [comment]
> @david-luna You'll also need to approve for merging, because the current last commit is mine.

We are in the same situation as #3662 so I cannot approve. I made a last commit & requested your review again like we did then

**Review by trentm — APPROVED** [review]
